docs/doxygen/pages.txt

```diff
@@ -51,7 +51,115 @@ The configurations for memory estimation are defined <a href="https://github.com
+These states are stored in the arrays @ref MQTTContext_t.outgoingPublishRecords and @ref MQTTContext_t.incomingPublishRecords;
+the length of each of these arrays can be configured with @ref MQTT_STATE_ARRAY_MAX_COUNT.
+This library does not store any subscription information, nor any information for QoS 0 publishes.
+
+When resuming a persistent session, the client library will resend PUBRELs for all PUBRECs had been received
+for incomplete outgoing QoS 2 publishes. If the broker does not resume the session, then all state information
+in the client will be reset.
+
+@note The library stores only the <i>state</i> of incomplete publishes and not the publishes themselves.
+If a persistent session is resumed, then @ref mqtt_publishtoresend_function should be called to obtain the
+packet identifiers of incomplete publishes, followed by a call to @ref mqtt_publish_function to resend the
+unacknowledged publish.
+
+@section mqtt_receivepackets Packet Reception
+
+MQTT Packets are received from the network with calls to @ref mqtt_processloop_function or @ref mqtt_receiveloop_function. The functions are mostly identical,
+with the exception of keep-alive; the former sends ping requests and processes ping responses to ensure the keep-alive interval does not expire, while the latter does not.
+Since calls to @ref mqtt_publish_function, @ref mqtt_subscribe_function, and @ref mqtt_unsubscribe_function only send packets and do not
+wait for acknowledgments, a call to @ref mqtt_processloop_function or @ref mqtt_receiveloop_function must follow in order to receive any expected acknowledgment.
+The exception is @ref mqtt_connect_function; since a connection cannot be considered established until the server accepts a CONNECT packet with a CONNACK,
+the function waits until the CONNACK is received.
+
+@subsection mqtt_receivetimeout Timeouts
+@ref mqtt_connect_function, @ref mqtt_processloop_function, and @ref mqtt_receiveloop_function all accept timeout parameters for packet reception.
+For the connect function, if this value is set to 0, then instead of a time-based loop, it will attempt to call the transport receive function up
+to a fixed maximum number of times, which is defined by @ref MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT.
+
+For the loop functions, the timeout value represents the <i>minimum</i> duration that will be spent in the function, provided there are no network errors.
+Should the timeout be set to 0, then the loop will run for a single iteration. A single iteration of a loop consists of an attempt to receive a single byte from the network,
+followed by the reception of the entire packet if the receive was successful, and finally deserialization of the packet received and a call to the application callback. If the first read did not succeed,
+then instead the library checks if a ping request needs to be sent (only for the process loop).
```

Comment:
I made some plantUML diagrams for the operation of the loops:

MQTT_Connect():
```
@startuml

start
: Send CONNECT packet;
: count = 0;

repeat
	: Receive single byte;
repeat while (No network data AND \ncount < MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT) is (yes)
-> no or timeout == 0;

: Get rest of CONNACK packet;
: Deserialize CONNACK packet;
stop

@enduml
```

Process Loop:
```
@startuml

start

repeat
	: Receive single byte;
    if( success ) then (yes)
     	: Get rest of packet;
    	: Deserialize packet;
    else (no)
    	: Send ping request;
    endif

repeat while (**timeout** reached) is (no)
-> yes or timeout == 0;

stop

@enduml
```

ReceiveLoop:
```
@startuml

start

repeat
	: Receive single byte;
    if( success ) then (yes)
     	: Get rest of packet;
    	: Deserialize packet;
    else (no)
    endif

repeat while (**timeout** reached) is (no)
-> yes or timeout == 0;

stop

@enduml
```